### PR TITLE
Do not export TrieDB and TrieDBMut

### PR DESCRIPTION
**Pull request by sgkim126**
This patch makes codechain use Trie and TrieMut which are abstracted
interfaces instead of concrete implementations.

### CODE REVIEW

**Comment by Remagpie** [comment]
Nice patch! I totally forgot that we had `TrieFactory`.

File: util/merkle/benches/triedb.rs [hunk 1]
```diff
@@ -139,9 +139,11 @@ fn bench_read_multiple(b: &mut Bencher) {
 fn bench_write_single(b: &mut Bencher) {
     let mut db = TestDB::new(DB_SIZE);
     b.iter(|| {
-        let mut trie = db.trie_mut();
-        let item = random::<usize>() % DB_SIZE + DB_SIZE;
-        let _ = trie.insert(&item.to_be_bytes(), &item.to_be_bytes()).unwrap();
+        {
```
Comment (Remagpie) [inline util/merkle/benches/triedb.rs]: Why did you add a new block?

File: util/merkle/src/triedb.rs [hunk 2]
```diff
@@ -40,12 +40,12 @@ use crate::{Trie, TrieError};
...
 /// let mut root = H256::new();
-/// TrieDBMut::new(&mut memdb, &mut root).insert(b"foo", b"bar").unwrap();
-/// let t = TrieDB::try_new(&memdb, &root).unwrap();
+/// TrieFactory::create(&mut memdb, &mut root).insert(b"foo", b"bar").unwrap();
+/// let t = TrieFactory::readonly(&memdb, &root).unwrap();
 /// assert!(t.contains(b"foo").unwrap());
 /// assert_eq!(t.get(b"foo").unwrap().unwrap(), b"bar".to_vec());
 /// ```
-pub struct TrieDB<'db> {
+pub(crate) struct TrieDB<'db> {
```
Comment (Remagpie) [inline util/merkle/src/triedb.rs]: Is this required?

File: util/merkle/benches/triedb.rs [hunk 3]
```diff
@@ -92,14 +92,14 @@ impl TestDB {
         }
     }
 
-    fn trie(&self) -> TrieDB {
+    fn trie<'db>(&'db self) -> Box<dyn Trie + 'db> {
```
Comment (Remagpie) [inline util/merkle/benches/triedb.rs]: The return type became a lot verbose than before, and it seems it's coming from the implementation of `TrieFactory`'s methods.
I suggest simplifying this by refactoring `TrieFactory` and remove those `Box` and lifetimes if possible.

Comment (sgkim126) [reply util/merkle/benches/triedb.rs]: You cannot remove Box because the size of the return type should be known at the compile time.

Comment (Remagpie) [reply util/merkle/benches/triedb.rs]: Oh, I see.

Comment (sgkim126) [reply util/merkle/benches/triedb.rs]: Ah, I think `impl Trait` can work. I'll try it.

File: util/merkle/benches/triedb.rs (hunk 1 above)
Comment (sgkim126) [reply util/merkle/benches/triedb.rs]: To drop trie before flushing db.

Comment (sgkim126) [reply util/merkle/benches/triedb.rs]: Both `trie_mut` and `flush` mutably borrow `db`.

Comment (Remagpie) [reply util/merkle/benches/triedb.rs]: It seems it's not related to the commit title "Do not export TrieDB and TrieDBMut".
Could you extract it as a new commit?

File: util/merkle/src/triedb.rs (hunk 2 above)
Comment (sgkim126) [reply util/merkle/src/triedb.rs]: Yes. The TrieDB should not be exported out of the crate, but it's used inside the crate.

File: util/merkle/benches/triedb.rs (hunk 1 above)
Comment (sgkim126) [reply util/merkle/benches/triedb.rs]: It's related. The return type of `trie_mut` was not bounded to the lifetime of `db`, but this patch makes it bounded.

File: util/merkle/src/triedb.rs (hunk 2 above)
Comment (Remagpie) [reply util/merkle/src/triedb.rs]: Isn't it exported only when we `pub use` in the `mod.rs`?
I think removing `pub` in `mod.rs` is enough.

File: util/merkle/benches/triedb.rs (hunk 1 above)
Comment (Remagpie) [reply util/merkle/benches/triedb.rs]: Oh, I thought that there would be no compile error even if we flatten the block.
It seems we need it.

File: util/merkle/src/triedb.rs (hunk 2 above)
Comment (sgkim126) [reply util/merkle/src/triedb.rs]: I want to guarantee that this doesn't go out of the crate.

File: util/merkle/benches/triedb.rs (hunk 3 above)
Comment (sgkim126) [reply util/merkle/benches/triedb.rs]: `impl Trait` worked.

File: util/merkle/src/triedb.rs (hunk 2 above)
Comment (Remagpie) [reply util/merkle/src/triedb.rs]: Hmm. I think that we didn't "have to" introduce this change.
You can leave it as is if you want, but I'd like to suggest minimizing the change if possible.
